txt.1
.\"    Automatically generated by mango(1)
.TH "txt" 1 "2014-01-14" "version 2014-01-14" "User Commands"
.SH "NAME"
txt \- Command txt is a templating language for shell programming.
.SH "SYNOPSIS"
.B txt
.RB [ \-left
.IR Left ]
.RB [ \-right
.IR Right ]
.RB [ \-template
.IR Template ]
.RB [ \-e
.IR Expression ]
.RB [ \-html ]
.RB [ \-R
.IR RecordSeparator ]
.RB [ \-F
.IR FieldSeparator ]
.RB [ \-L
.IR LinePattern ]
.RB [ \-json ]
.RB [ \-csv ]
.RB [ \-no-stdin ]
.RB [ \-header
.IR Header ]
.B template\-files*
.SH "DESCRIPTION"
Input 
.PP
The input to the template comes from stdin. 
It is parsed in one of five ways. 
.PP
The default is to split stdin into records and fields, using the 
.B \-R
and 
.B \-F
flags respectively, similar to awk(1), and dot is set to a list of records (see below). 
If header is set, dot is a list of maps with the specified names as keys. 
.PP
If the 
.B \-L
flag is specified stdin is broken into records as with the default, but the fields are defined by the capture groups of the regular expression 
.B \-L.
Records that do not match 
.B \-L
are skipped. 
If 
.B \-L
contains named capture groups each record is a dictionary of only the named captures' values for that record. 
Otherwise, dot is a list of records (see below) of the capture groups' values. 
If header is set, dot is a list of maps with the specified names as keys, overriding any names from capture groups. 
.PP
If the 
.B \-csv
flag is specified, stdin is treated as a CSV file, as recognized by the encoding/csv package. 
If the 
.B \-header
flag is not specified, the first record is used as the header. 
Dot is set to a list of maps, with the header for each column as the key. 
.PP
If the 
.B \-json
flag is specified, stdin is treated as JSON. 
Dot is set as the decoded JSON. 
.PP
If the 
.B \-no-stdin
flag is specified, stdin is not read. 
Dot is not set. 
.PP
Records 
.PP
When using 
.B \-F
or 
.B \-L
without a header, or in the case of 
.B \-L
without named capture groups, dot is a list of records. 
.PP
Each record has two fields, Fields and Line. 
Line is the complete unaltered input of that record. 
Fields are the values of each field in that record. 
If dot is a record 
.PP
.RS
{{.}}
.RE
.PP
is the same as 
.PP
.RS
{{.Line}}
.RE
.PP
Records have a method F that takes an integer n and returns the nth field if it exists and the empty string otherwise. 
If n is negative it returns the (n\-1)th field from the end. 
.PP
If n is positive and the nth field exists, then 
.PP
.RS
{{.F n}}
.RE
.PP
is equivalent to 
.PP
.RS
{{index \&. n}}
.sp 0
.sp
.RE
.PP
Templates 
.PP
The templating language is documented at http://golang.org/pkg/text/template with the single difference that if the first line at the top of the file begins with #! 
that line is skipped. 
If the 
.B \-html
flag is used, escaping functions are automatically added to all outputs based on context. 
.PP
Any command line arguments after the flags are treated as filenames of templates. 
The templates are named after the basename of the respective filename. 
The first file listed is the main template, unless the 
.B \-template
flag specifies otherwise. 
If the 
.B \-e
flag is used to define an inline template, it is always the main template, and the 
.B \-template
flag is illegal. 
.PP
Regular Expressions 
.PP
All regular expressions are RE2 regular expression with the Perl syntax and semantics. 
The syntax is documented at http://golang.org/pkg/regexp/syntax/#hdr\-Syntax 
.PP
Functions 
.PP
Built in functions are documented at http://golang.org/pkg/text/template#hdr\-Functions 
.PP
The following additional functions are defined: 
.PP
.RS
slice what start stop
.sp 0
.RS
slices what from start to stop.
.sp 0
stop is optional. what must be a list or string.
.sp 0
start or stop may be negative. If start or stop exceed the bounds
.sp 0
of what, the largest slice of what that exists is returned.
.sp 0
.sp
.RE
nl string
.sp 0
.RS
append a newline to the end of string, if it does not end in newline.
.sp 0
.sp
.RE
parseCSV headerspec filename
.sp 0
.RS
headerspec is a comma\-separated list of headers or "" to use the headers
.sp 0
in filename.
.sp 0
Dot is set to the contents of the CSV file as with \-csv.
.sp 0
If the file cannot be opened or its contents are malformed, execution
.sp 0
stops.
.sp 0
.sp
.RE
parseJSON filename
.sp 0
.RS
Read the JSON encoded file into dot or halt execution if decoding fails
.sp 0
or the file cannot be opened.
.sp 0
Dot is set to the contents of the JSON file as with \-json.
.sp 0
.sp
.RE
parseLine header FS LP filename
.sp 0
.RS
Read filename with line pattern splitting as specified by the RS and
.sp 0
LP regular expressions, and an optional header header.
.sp 0
If header is not "", the names in header will be used as the field names.
.sp 0
If RS or LP are "", the respective value of \-R or \-L is used.
.sp 0
.sp
.RE
parse header RS FS filename
.sp 0
.RS
Read filename with the default record and file splitting as specified
.sp 0
by the RS and FS regular expressions, and optional header header.
.sp 0
If header is not "", the names in header will be used as the field names.
.sp 0
If RS or FS are "", the respective value of \-R or \-F is used.
.sp 0
.sp
.RE
read filename
.sp 0
.RS
Read filename completely as a single string.
.sp 0
Execution halts if the file cannot be read.
.sp 0
.sp
.RE
quoteCSV string
.sp 0
.RS
Apply the appropriate CSV quoting rules to string.
.sp 0
.sp
.RE
toJSON what
.sp 0
.RS
Encode what as JSON. Execution halts if
.sp 0
http://golang.org/pkg/encoding/json/#Marshal errors.
.sp 0
.sp
.RE
equalFold string\-one string\-two
.sp 0
.RS
Reports whether the UTF\-8 encoded string\-one and string\-two are equal
.sp 0
under Unicode case\-folding.
.sp 0
.sp
.RE
fields string
.sp 0
.RS
Split string around whitespace.
.sp 0
.sp
.RE
join separator strings
.sp 0
.RS
Join the list in strings by the string separator.
.sp 0
.sp
.RE
lower string
.sp 0
.RS
Lowercase string.
.sp 0
.sp
.RE
upper string
.sp 0
.RS
Uppercase string.
.sp 0
.sp
.RE
title string
.sp 0
.RS
Titlecase string.
.sp 0
.sp
.RE
trimCutset cutset string
.sp 0
.RS
Return string with all leading and trailing runes in cutset removed.
.sp 0
.sp
.RE
trimLeft cutset string
.sp 0
.RS
Return string with all leading runes in cutset removed.
.sp 0
.sp
.RE
trimRight cutset string
.sp 0
.RS
Return string with all trailing runes in cutset removed.
.sp 0
.sp
.RE
trimPrefix prefix string
.sp 0
.RS
Return string with prefix removed.
.sp 0
.sp
.RE
trimSuffix suffix string
.sp 0
.RS
Return string with suffix removed.
.sp 0
.sp
.RE
trim string
.sp 0
.RS
Return string with all leading and trailing whitespace removed.
.sp 0
.sp
.RE
quoteGo string
.sp 0
.RS
Return string quoted as a Go string literal. Escapes non\-printable
.sp 0
runes. Should work for most languages that accept UTF\-8 source.
.sp 0
.sp
.RE
quoteGoASCII string
.sp 0
.RS
As quoteGo except any non\-ASCII runes are escaped to hexcodes.
.sp 0
Should work for most languages.
.sp 0
.sp
.RE
match pattern string
.sp 0
.RS
Return whether string matches the regex in pattern.
.sp 0
Execution halts if pattern is not a valid regular expression.
.sp 0
.sp
.RE
find pattern string
.sp 0
.RS
Returns all substrings of string that match pattern.
.sp 0
Execution halts if pattern is not a valid regular expression.
.sp 0
.sp
.RE
replace pattern spec string
.sp 0
.RS
Replace all substrings in string matching pattern by spec.
.sp 0
Execution halts if pattern is not a valid regular expression.
.sp 0
.sp
.RE
split pattern string
.sp 0
.RS
Split string into a list of substrings separated by pattern.
.sp 0
Execution halts if pattern is not a valid regular expression.
.sp 0
.sp
.RE
env key
.sp 0
.RS
Returns the environment variable key or "".
.sp 0
.sp
.RE
exec name args*
.sp 0
.RS
Execute command name with args. Stdin is nil.
.sp 0
Stderr shares the stderr of txt(1).
.sp 0
Stdout is returned as a string.
.sp 0
.sp
.RE
pipe name args* input
.sp 0
.RS
Execute command name with args with input as stdin.
.sp 0
Otherwise, like exec.
.RE
.RE
.SH "OPTIONS"
.TP
.BR "\-left " Left " = {{"
set left template delimiter 
.TP
.BR "\-right " Right " = }}"
set right template delimiter 
.TP
.BR "\-template " Template
which template to invoke, otherwise first listed 
.TP
.BR "\-e " Expression
expression to use as main template 
.TP
.BR "\-html "
use html\-aware automatic escaping against code injection 
.TP
.BR "\-R " RecordSeparator
record separator 
.TP
.BR "\-F " FieldSeparator
field separator 
.TP
.BR "\-L " LinePattern
line pattern, regex must contain capture groups 
.TP
.BR "\-json "
treat input as JSON 
.TP
.BR "\-csv "
treat input as CSV 
.TP
.BR "\-no-stdin "
do not read stdin 
.TP
.BR "\-header " Header
specify a header as a comma\-separated list 


txt.go
package main

import (
	"flag"
	"log"
	"os"
)

const (
	RS = "\n+"
	FS = `\s+`
)

var (
	Left  = flag.String("left", "{{", "set left template delimiter")
	Right = flag.String("right", "}}", "set right template delimiter")

	Template   = flag.String("template", "", "which template to invoke, otherwise first listed")
	Expression = flag.String("e", "", "expression to use as main template")

	Html = flag.Bool("html", false, "use html-aware automatic escaping against code injection")

	RecordSeparator = flag.String("R", RS, "record separator")
	FieldSeparator  = flag.String("F", FS, "field separator")
	LinePattern     = flag.String("L", "", "line pattern, regex must contain capture groups")

	Json    = flag.Bool("json", false, "treat input as JSON")
	Csv     = flag.Bool("csv", false, "treat input as CSV")
	NoStdin = flag.Bool("no-stdin", false, "do not read stdin")

	Header = flag.String("header", "", "specify a header as a comma-separated list")
)

//Usage: %name %flags template-files*
func main() {
	log.SetFlags(0)

	flag.Usage = func() {
		log.Printf("Usage: %s [-json|-csv|-no-stdin] -html -left=delim -right=delim\n", os.Args[0])
		p := log.Println
		p("\t[-e=template|-template=name] -R=RE [-F=RE|-L=RE]")
		p("\t-header=headerspec template-files*")

		p(" Template control:")
		p("  -left delim:    set the left delimiter in templates")
		p("  -right delim:   set the right delimiter in templates")
		p("  -html:          use html-aware autoescaping")
		p(" Template selection:")
		p("  -e template:    specifiy main template as string")
		p("  -template file: say which of the template files is the main template")
		p(" Input handling")
		p("  -json:          parse input as JSON")
		p("  -csv:           parse input as CSV")
		p("  -no-stdin:      do not read stdin")
		p("  -R regex:       record separator, defaults to \"\\n+\"")
		p("  -F regex:       field separator, defaults to \"\\s+\"")
		p("  -L regex:       line-matching pattern")
		p("  -header list:   comma-separated list of field names")

		p("-e and -template are mutually exclusive")
		p("Only one of -json, -csv, -no-stdin, -F, or -L can be specified")
		p("-header can only be used with -csv, -F, or -L")
		p("-R can only be used with -F or -L")

		os.Exit(2)
	}
	flag.Parse()

	//validate arguments
	fail := false
	if *Expression != "" && *Template != "" {
		fail = true
	}
	if multiple(*Csv, *Json, *NoStdin) {
		fail = true
	}
	notregex := oneOf(*Csv, *Json, *NoStdin)
	if notregex && *RecordSeparator != RS {
		fail = true
	}
	if notregex && *FieldSeparator != FS {
		fail = true
	}
	if notregex && *LinePattern != "" {
		fail = true
	}
	if (*Json || *NoStdin) && *Header != "" {
		fail = true
	}
	if *FieldSeparator != FS && *LinePattern != "" {
		fail = true
	}
	if fail {
		log.Println("Invalid combination of flags")
		flag.Usage()
		os.Exit(2)
	}
	args := flag.Args()

	var which string

	//parse templates

	//If -e used, use as main template, even if other templates specified.
	//If template(s) specified use first as main unless specified by flag.
	//otherwise no template
	if len(args) > 0 {
		which = args[0]
		if *Template != "" {
			which = *Template
		}
	} else if *Expression == "" {
		log.Fatalln("No template(s) specified")
	}
	tmpl, err := Parse(*Html, *Expression, *Left, *Right, funcs, args...)
	if err != nil {
		log.Fatalln(err)
	}

	hdr := splitHeader(*Header)

	//parse input
	var stdin interface{}
	if *Csv {
		stdin, err = CSV(hdr, os.Stdin)
	} else if *Json {
		stdin, err = JSON(os.Stdin)
	} else if *LinePattern != "" {
		stdin, err = SubmatchSplit(hdr, *RecordSeparator, *LinePattern, os.Stdin)
	} else if !*NoStdin {
		stdin, err = Split(hdr, *RecordSeparator, *FieldSeparator, os.Stdin)
	}
	if err != nil {
		log.Fatalln(err)
	}

	//run program
	if err = tmpl.ExecuteTemplate(os.Stdout, which, stdin); err != nil {
		log.Fatalln(err)
	}
}
